### DISCUSSION TITLE
Chocolatey Products Alpha/Beta Releases

### DISCUSSION THREAD

**Discussion by gep13** [chocolatey/choco#2995] [General] [answered]
This discussion is being used to capture any issues related to the usage of the alpha versions for the various Chocolatey Components, that are going to support NuGet v3 feeds, as well as .NET Framework 4.8.

- Chocolatey CLI
  - [2.0.0-beta-20230426](https://community.chocolatey.org/packages/chocolatey/2.0.0-beta-20230426) - [Release Notes](https://docs.chocolatey.org/en-us/choco/release-notes#beta-20230426-april-26-2023)
  - [2.0.0-beta-20230412](https://community.chocolatey.org/packages/chocolatey/2.0.0-beta-20230412) - [Release Notes](https://docs.chocolatey.org/en-us/choco/release-notes#beta-20230412-april-12-2023)
  - [2.0.0-beta-20230321](https://community.chocolatey.org/packages/chocolatey/2.0.0-beta-20230321) - [Release Notes](https://docs.chocolatey.org/en-us/choco/release-notes#beta-20230321-march-21-2023)
  - [2.0.0-alpha-20230307](https://community.chocolatey.org/packages/chocolatey/2.0.0-alpha-20230307) - [Release Notes](https://docs.chocolatey.org/en-us/choco/release-notes#alpha-20230307-march-7-2023)
  - [2.0.0-alpha-20230221](https://community.chocolatey.org/packages/chocolatey/2.0.0-alpha-20230221) - [Release Notes](https://docs.chocolatey.org/en-us/choco/release-notes#alpha-20230221-february-21-2023)
  - [2.0.0-alpha-20230124](https://community.chocolatey.org/packages/chocolatey/2.0.0-alpha-20230124) - [Release Notes](https://docs.chocolatey.org/en-us/choco/release-notes#alpha-20230124-january-24-2023)
- Chocolatey GUI
  - [2.0.0-beta-20230426](https://community.chocolatey.org/packages/chocolateygui/2.0.0-beta-20230426) - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui/release-notes#beta-20230426-april-26-2023)
  - [2.0.0-beta-20230412](https://community.chocolatey.org/packages/chocolateygui/2.0.0-beta-20230412) - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui/release-notes#beta-20230412-april-12-2023)
  - [2.0.0-beta-20230321](https://community.chocolatey.org/packages/chocolateygui/2.0.0-beta-20230321) - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui/release-notes#beta-20230321-march-21-2023)
  - [2.0.0-alpha-20230307](https://community.chocolatey.org/packages/chocolateygui/2.0.0-alpha-20230307) - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui/release-notes#alpha-20230307-march-7-2023)
  - [2.0.0-alpha-20230221](https://community.chocolatey.org/packages/ChocolateyGUI/2.0.0-alpha-20230221) - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui/release-notes#alpha-20230221-february-21-2023)
- Chocolatey Licensed Extension
  - 6.0.0-beta-20230426 - Contact support for download/installation instruction - [Release Notes](https://docs.chocolatey.org/en-us/licensed-extension/release-notes#beta-20230426-april-26-2023)
  - 6.0.0-beta-20230412 - Contact support for download/installation instruction - [Release Notes](https://docs.chocolatey.org/en-us/licensed-extension/release-notes#beta-20230412-april-12-2023)
  - 6.0.0-beta-20230321 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/licensed-extension/release-notes#beta-20230321-march-21-2023)
  - 6.0.0-alpha-20230307 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/licensed-extension/release-notes#alpha-20230307-march-7-2023)
  - 6.0.0-alpha-20230221 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/licensed-extension/release-notes#alpha-20230221-february-21-2023)
- Chocolatey GUI Licensed Extension
  - 2.0.0-beta-20230426 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui-licensed-extension/release-notes#beta-20230426-april-26-2023)
  - 2.0.0-beta-20230412 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui-licensed-extension/release-notes#beta-20230412-april-12-2023)
  - 2.0.0-beta-20230321 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui-licensed-extension/release-notes#beta-20230321-march-21-2023)
  - 2.0.0-alpha-20230221 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/chocolatey-gui-licensed-extension/release-notes#alpha-20230221-february-21-2023)
- Chocolatey Agent
  - 2.0.0-beta-20230426 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/agent/release-notes#beta-20230426-april-26-2023)
  - 2.0.0-beta-20230412 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/agent/release-notes#beta-20230412-april-12-2023)
  - 2.0.0-beta-20230321 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/agent/release-notes#beta-20230321-march-21-2023)
  - 2.0.0-alpha-20230221 - Contact support for download/installation instructions - [Release Notes](https://docs.chocolatey.org/en-us/agent/release-notes#alpha-20230221-february-21-2023)

If you find an issue with any of these pre-release packages, please first check the known issues list below, and if it is not already covered, add an answer to the discussion below and we can discuss further.

# Known Issues

The following lists, which will be added to as any new issues are discovered, are of known issues with the pre-release version of Chocolatey CLI, Chocolatey GUI, Chocolatey Licensed Extension, Chocolatey GUI Licensed Extension, and Chocolatey Agent.

This is not intended to provide full details on the issues, but if more information is required to make things clear, please reach out and we can provide additional information.

> **Note** 
>
> Items that have been scored out, i.e. ~~like this~~ have been resolved in a newer package version

## Chocolatey CLI

- The Chocolatey Community Repository currently does not have a NuGet v3 feed available, so to test this new functionality you will need to use _something_ else.  For example, a Nexus Repository, Azure Artifacts etc.
- There have been breaking changes to Chocolatey's Public API exposed in Chocolatey.Lib.  Anyone directly using Chocolatey.Lib in a custom application will likely have work to do to adopt the changes.

### 2.0.0-beta-20230426

- Attempting to do something like `choco list --source cygwin` will not list the applications that are currently installed via cygwin. This came about due to the change to make `choco list` only return currently installed packages.  We are going to carry this as a known issue into the 2.0.0 release, and a new feature in a future version of Chocolatey CLI will include the ability to do something like `choco list --include-alternative-sources` in the same way as you can do `choco list --include-programs` for listing applications installed directly in add/remove programs
- When searching for packages, filtering of the results based on criteria that is only supported on Chocolatey Community Repository (for example, --approved-only) is currently not supported in the oData query.  Instead, this filtering is performed on the client side, within Chocolatey CLI.  As a result, packages returned from the server may be filtered out within Chocolatey CLI, meaning that you may not get full pages of resulted returend.
- When using the `--force-dependencies` option on the `choco uninstall` command, an error is reported when a dependency constraint is violated, rather than a warning and leaving it in place, while uninstalling the other packages/dependencies
- ~~Running the command `choco push -h` doesn't work as expected, i.e. doesn't show the help output, unless you have a `defaultPushSource` configured~~
- ~~Executing the command `choco list -li` does not return currently installed applications from add/remove programs~~
- ~~Using `--page-size` and `--page` can have unexpected results when targetting a v3 feed~~

### 2.0.0-beta-20230412

- Running the command `choco push -h` doesn't work as expected, i.e. doesn't show the help output, unless you have a `defaultPushSource` configured
- Executing the command `choco list -li` does not return currently installed applications from add/remove programs
- Using `--page-size` and `--page` can have unexpected results when targetting a v3 feed
- When searching for packages, ordering of the results is currently not respected, i.e. using `--approved-only`, etc.
- When using the `--force-dependencies` option on the `choco uninstall` command, an error is reported when a dependency constraint is violated, rather than a warning and leaving it in place, while uninstalling the other packages/dependencies

### 2.0.0-beta-20230321

- Executing the command `choco list -li` does not return currently installed applications from add/remove programs
- A priority value attributed to a source is not currently respected
- Using `--page-size` and `--page` can have unexpected results when targetting a v3 feed
- When searching for packages, ordering of the results is currently not respected, i.e. using `--approved-only`, etc.
- When using the `--force-dependencies` option on the `choco uninstall` command, an error is reported when a dependency constraint is violated, rather than a warning and leaving it in place, while uninstalling the other packages/dependencies

### 2.0.0-alpha-20230307

- ~~> The source failed to get a NuGet.Protocol.Core.Types.ListResource resource~~
  - ~~This is discussed further [below](https://github.com/chocolatey/choco/discussions/2995#discussioncomment-5242622) where a workaround is provided~~
- A priority value attributed to a source is not currently respected
- Using `--page-size` and `--page` can have unexpected results when targetting a v3 feed
- When searching for packages, ordering of the results is currently not respected, i.e. using `--approved-only`, etc.
- When using the `--force-dependencies` option on the `choco uninstall` command, an error is reported when a dependency constraint is violated, rather than a warning and leaving it in place, while uninstalling the other packages/dependencies

### 2.0.0-alpha-20230221

- ~~We have seen some intermittent issues with the resolving of, and installation of, package dependencies.~~
- A priority value attributed to a source is not currently respected
- Using `--page-size` and `--page` can have unexpected results when targetting a v3 feed
- When searching for packages, ordering of the results is currently not respected, i.e. using `--approved-only`, etc.
- When using the `--force-dependencies` option on the `choco uninstall` command, an error is reported when a dependency constraint is violated, rather than a warning and leaving it in place, while uninstalling the other packages/dependencies
- ~~The `--disable-repository-optimizations` option does not change the outgoing oData queries when doing Chocolatey commands~~
- ~~There are problems with query results when attempting to use the `--version` option~~

### 2.0.0-alpha-20230124

- ~~Pushing packages to the Chocolatey Community Repository incorrectly states that a package was pushed successfully when actually no package was pushed.~~
- ~~The `choco push` command does not correctly respond to error codes returned from a package repository.~~
- ~~When running the `choco pack` command, custom validations of Chocolatey specific elements in the nuspec are not currently working.~~
- ~~`--page-size` and `--page` options are not supported when targeting v2 feeds~~
- ~~Unable to re-install a package due to temporary files left after initial installation~~
- ~~Certain Chocolatey commands can result in errors due to long file path problems~~
- ~~When uninstalling a package, package dependencies are incorrectly attempted to also be removed~~
- ~~Prerelease dependencies are installed when not requested - as detailed [here](https://github.com/chocolatey/choco/discussions/2995#discussioncomment-4877028)~~
- ~~We have seen some intermittent issues with the resolving of, and installation of, package dependencies.~~
- A priority value attributed to a source is not currently respected
- Using `--page-size` and `--page` can have unexpected results
- When searching for packages, ordering of the results is currently not respected, i.e. using `--order-by-popularity`
  - This has been partially resolved.  The `order-by-popularity` option is now correctly respected.

### 2.0.0-beta-20230412

- For some repository servers, for example, Sonatype Nexus, no paging information is displayed when accessing a v3 feed - it will always show all packages from the feed in a single page.  Other repository servers are known to function correctly.

### 2.0.0-beta-20230412

- No paging information is displayed when accessing a v3 feed - it will always show all packages from the feed in a single page

### 2.0.0-beta-20230321

- No paging information is displayed when accessing a v3 feed - it will always show all packages from the feed in a single page

### 2.0.0-alpha-20230307

- No paging information is displayed when accessing a v3 feed - it will always show all packages from the feed in a single page
- ~~There are currently no localisations (translations into other languages).~~

### 2.0.0-alpha-20230221

- No paging information is displayed when accessing a v3 feed - it will always show all packages from the feed in a single page
- There are currently no localisations (translations into other languages).
- ~~The Chocolatey GUI MSI includes the chocolatey.dll and chocolatey.pdb file, when it shouldn't~~

### 6.0.0-beta-20230426

- There are currently no known issues with this package, other than those which are already present in Chocolatey CLI.
- ~~There is currently an issue with running `choco uninstall` on a package created by the `choco sync` command - It will uninstall the package and the application, but it will leave behind the `lib` directory with some nuspec files.~~

### 6.0.0-beta-20230412

- There is currently an issue with running `choco uninstall` on a package created by the `choco sync` command - It will uninstall the package and the application, but it will leave behind the `lib` directory with some nuspec files.

### 6.0.0-beta-20230321

- ~~In this pre-release version, Intune support is not current available.~~

### 6.0.0-alpha-20230307

- In this pre-release version, Intune support is not currently available.

### 6.0.0-alpha-20230221

- In this pre-release version, Intune support is not currently available.

### 2.0.0-beta-20230426

- There are currently no known issues with this package, other than those which are already present in Chocolatey GUI.

### 2.0.0-beta-20230412

- There are currently no known issues with this package, other than those which are already present in Chocolatey GUI.

### 2.0.0-beta-20230321

- There are currently no known issues with this package, other than those which are already present in Chocolatey GUI.

### 2.0.0-alpha-20230221

- There are currently no known issues with this package, other than those which are already present in Chocolatey GUI.

### 2.0.0-beta-20230426

- There are currently no known issues with this package.

### 2.0.0-beta-20230412

- There are currently no known issues with this package.

### 2.0.0-beta-20230321

- There are currently no known issues with this package.

### 2.0.0-alpha-20230221

- There are currently no known issues with this package.

**Comment by TheCakeIsNaOH** [comment]
If the absolute latest version of a not installed dependency is prerelease, the prerelease version will be installed, even if there is a release version available that would satisfy the version contraints.

Example: `.\choco install synctrayzor -n` on the alpha will install `dotnetfx` `v4.8.1-rtw` instead of `4.8.0.20220524` (assuming that `dotnetfx` is not already installed).

**Reply by gep13** [reply]
@TheCakeIsNaOH thanks for bringing this up. During the last sprint of work, this issue has been addressed, and should be included within the next release.

**Comment by Sainan** [comment]
2.0.0-alpha-20230307 seems to have broken installation from v3 sources, it just says "The source ... failed to get a NuGet.Protocol.Core.Types.ListResource resource"

Not entirely sure what that means, but the source in question supports everything that is specified as "required" in NuGet V3 API documentation: https://learn.microsoft.com/en-us/nuget/api/overview#resources-and-schema

Also works fine with 2.0.0-alpha-20230221 and 2.0.0-alpha-20230124.

**Reply by gep13** [reply]
Can you provide a few more details here?

Like what V3 source are you using? Is it something like Azure Artifacts?  Or Nexus, or Artifactory.

If you could provide some log files as well, that would be great, thank you!

**Reply by Sainan** [reply]
The source is `https://choco.calamity.gg/index.json` (https://github.com/calamity-inc/choco.calamity.gg).

Also, I know that this source does not list a SearchQueryService endpoint, but providing one made no difference as to whether this error shows up in my testing. The only way to fix it was to downgrade to 2.0.0-alpha-20230221 or 2.0.0-alpha-20230124.

Specifically, I was trying to install [calamity-inc/chocogen](https://github.com/calamity-inc/chocogen) like so:
```
choco source add -n "Calamity, Inc." -s https://choco.calamity.gg/index.json
choco install chocogen
```

Unfortunately, the log does not provide more information than what I've already said:
> 2023-03-08 15:08:05,443 38344 [WARN ] - The source https://choco.calamity.gg/index.json failed to get a NuGet.Protocol.Core.Types.ListResource resource

**Reply by TheCakeIsNaOH** [reply]
The issue is here: https://github.com/chocolatey/choco/commit/42df644d04f5a1a51fdad710a48835b2a2b88f89#diff-e1f39f082ffc1bb9680326d3576b2dd1c175d3e3af3d74064a5fd700afedbd02R345-R347

The list resource is not required for v3 feeds, and they do not have a native list resource, instead, the list resources for v3 feeds point to the v2 list endpoint (if available).  So the list resource cannot be assumed to be available, the availability needs to be checked.

**Reply by Sainan** [reply]
Thanks for the insight, cake! I would highly appreciate if Chocolatey 2.0.0 would support pure V3 servers since V2 has no documentation so I can't implement it for our package repository.

**Reply by gep13** [reply]
@Sainan what happens if you try:

```
choco install chocogen --disable-repository-optimizations
```

**Reply by gep13** [reply]
I have just tested the suggestion above, and I can confirm that it works around the problem that you are seeing here.  

For the time being until we can get this fixed up, I would suggest that you either revert back to version [2.0.0-alpha-20230221](https://community.chocolatey.org/packages/chocolatey/2.0.0-alpha-20230221) of the Chocolatey alpha, or you run:

```
choco feature disable --name="'usePackageRepositoryOptimizations'"
```

to disable the optimization feature that is currently causing you a problem.

**Reply by gep13** [reply]
A fix has been put in place for this problem, and it will be available in the next release of the Chocolatey CLI.

**Comment by maxihafer** [comment]


**Reply by TheCakeIsNaOH** [reply]
@MaxiHafer you are seeing an issue with upstream Mono: https://github.com/mono/mono/issues/12406

Issue #2313 was relating to using `nuget.exe` while building Chocolatey CLI, not usage of Chocolatey CLI.

**Reply by pauby** [reply]
I'm a little confused. There is no beta version of the Docker image.

**Reply by pauby** [reply]
It was mentioned:

> The linux docker image of v2.0.0-beta-20230321 fails to resolve SSL-Certificates

There is no Linux Docker image of v2.0.0-beta-20230321.

**Comment by nrk-akjaer** [comment]
I'm running into an issue related to upper/lowercase that I cannot replicate on 1.x. Tested on `2.0.0-beta-20230426`.

If the package name does not conform with the actual `id`, the behavior/output of `choco.exe upgrade --noop <package> -r` is different.

This is problematic because [Chocolatey Puppet Module](https://github.com/puppetlabs/puppetlabs-chocolatey) uses this to check for newer versions when a package is set to `latest`.

~~As all `id`'s are expected to be lowercase in the `nuspec`, would it not make sense for `choco` to convert the package name to lower-case?~~
^ Nevermind, this seems to be a bad idea. See **EDIT**.

_DISCLAIMER_: We're using `sleet` for a minimalistic v3 repository, and this is hosted on Linux as well (where the file system is case-sensitive) so not sure if that could have an impact. Just in case that would somehow matter here.

Using `7zip` as an example below, but have other packages where this also happens.

**When using _lowercase_**
```shell
> choco.exe upgrade --noop 7zip -r
7zip|22.1|22.1.0|false
```

**When using _UPPERCASE_**
```shell
> choco.exe upgrade --noop 7Zip -r
Chocolatey would have used NuGet to install packages (if they are not already installed):
7Zip

chocolatey-compatibility.extension v1.0.0
Progress: Downloading chocolatey-core.extension 1.4.0... 100%

chocolatey-core.extension v1.4.0

7zip.install v22.1.0
Would have run 'chocolateyInstall.ps1':
$ErrorActionPreference = 'Stop'

$toolsDir = "$(Split-Path -parent $MyInvocation.MyCommand.Definition)"
$filePath = if ((Get-OSArchitectureWidth 64) -and $env:chocolateyForceX86 -ne $true) {
       Write-Host "Installing 64 bit version" ; Get-Item $toolsDir\*_x64.exe }
else { Write-Host "Installing 32 bit version" ; Get-Item $toolsDir\*_x32.exe }

$packageArgs = @{
  packageName    = '7zip.install'
  fileType       = 'exe'
  softwareName   = '7-zip*'
  file           = $filePath
  silentArgs     = '/S'
  validExitCodes = @(0)
}
Install-ChocolateyInstallPackage @packageArgs
Remove-Item $toolsDir\*.exe -ea 0 -force

$installLocation = Get-AppInstallLocation $packageArgs.softwareName
if (!$installLocation)  { Write-Warning "Can't find 7zip install location"; return }
Write-Host "7zip installed to '$installLocation'"

Install-BinFile '7z' $installLocation\7z.exe

Progress: Downloading 7zip 22.1.0... 100%

7zip v22.1.0
```

**EDIT**:
- This seems to be the case in reverse as well, when a package `identifier` has (wrongfully?) used uppercase
- According to [Chocolatey documentation](https://docs.chocolatey.org/en-us/create/create-packages#naming-your-package) you require all packages to use lowercase `identifier`
- Microsoft's [nuspec documentation](https://learn.microsoft.com/en-us/nuget/reference/nuspec#id) specifies that `id` (aka identifier) is _case-insensitive_
- To add to the confusion, there was a change to the Puppet Module (see: https://github.com/puppetlabs/puppetlabs-chocolatey/commit/cb7f0e5e33ddef445af4a03b9ab339588a9dc698) which _forces_ `downcase` on the package name

Does `chocolatey` introduce it's own (more sane) requirements on top of the `nuspec`, or...? And should we blame the vendor who used upper-case `identifier`, the Puppet Module for forcing `downcase` or `choco` for not conforming to Microsoft's `nuspec` [standard](https://learn.microsoft.com/en-us/nuget/reference/nuspec#id)? 🤔

**Reply by gep13** [reply]
@nrk-akjaer thank you for taking the time to raise this, and apologies for not getting back to you about this sooner!

We have investigated this issue, and have been able to replicate the problem.  We have a fix in place, and this will be included in the next release of Chocolatey CLI.  https://github.com/chocolatey/choco/pull/3157

Thanks again!

**Reply by nrk-akjaer** [reply]
@gep13 I am super grateful! Thanks for all the hard work and effort you guys have done to bring the release of Chocolatey 2.0! ❤️

**Reply by pauby** [reply]
@nrk-akjaer At this particular time, I can't express how much we (as in me, the Engineering Team and the wider Chocolatey Team) appreciate those kind words. Thank you.

**Comment by Sainan** [comment]
Weirdly enough, 2.0.0-beta-20230321 as well as subsequent versions still exibit the V3 sources issue:
> The source https://choco.calamity.gg/index.json failed to get a NuGet.Protocol.Core.Types.ListResource resource

2.0.0-alpha-20230221 continues to work fine.

I should've probably validated this earlier, sorry.

**Reply by gep13** [reply]
@Sainan thank you for reaching out again.  I have tried to replicate the problem that you describe, but as of yet, I have not been able to.  Let we walk through the steps that I have used.

First up, I installed the alpha version of Chocolatey CLI, where this problem was first introduced. I then attempted the installation of the `chocogen` package that is on your feed, and it failed with the error that was reported:

```
choco install chocogen --source https://choco.calamity.gg/index.json
```

![image](https://github.com/chocolatey/choco/assets/1271146/1e1f1249-16b3-4d41-b69b-c292b9e80367)

After that, I upgraded to the latest version of Chocolatey CLI:

```
choco upgrade chocolatey --pre
```

and then ran the installation command again, and this time it worked...

![image](https://github.com/chocolatey/choco/assets/1271146/7c78f004-4bf9-4a1c-9b10-d90d63634ba6)

Can you please verify that the version of Chocolatey CLI that is reported in the output from the command, as shown above, is actually the current version of the beta version?

**Reply by gep13** [reply]
To further test this out, I stood up a Docker image for the baget container, which is known to also only have a v3 feed, and tested that, and it also seems to work as expected:

<img width="1416" alt="image" src="https://github.com/chocolatey/choco/assets/1271146/8d0da6c4-8ed8-496d-85a1-95b42a046cd7">

After standing up baget, I pushed the windirstat package to the feed that it created, and performed the choco install.

**Reply by gep13** [reply]
And then as a final sanity check, I installed the specific version of Chocolatey CLI that you mention, and I also can't replicate the problem that you mention `2.0.0-beta-20230321`:

<img width="1071" alt="image" src="https://github.com/chocolatey/choco/assets/1271146/59c1a3e6-5d0f-4591-a9e9-c0d66729f893">

**Reply by Sainan** [reply]
I see that you do the following, and that does indeed work fine (on a different note, will this also work fine with upgrades?):

```
choco install chocogen --source https://choco.calamity.gg/index.json
```

However, when I do the following, it fails:

```
choco source add -n "Calamity, Inc." -s https://choco.calamity.gg/index.json
choco install chocogen
```

**Reply by gep13** [reply]
I don't think it will fail with the same error as you previously reported though, right?  What error does it report?

I suspect the issue here is the source name.  Typically, a source name will not have any spaces or extra characters in it.

Can you try with just:

```
choco source add -n "Calamity, Inc." -s https://choco.calamity.gg/index.json
```

I am not sure I follow what you mean when you ask "on a different note, will this also work fine with upgrades?".  Can you elaborate on this?

**Reply by Sainan** [reply]
Assuming you mean I should do something like this:
```
choco source add -n "calamityinc" -s https://choco.calamity.gg/index.json
```
I do indeed get the same error:
> The source https://choco.calamity.gg/index.json failed to get a NuGet.Protocol.Core.Types.ListResource resource

---

For the upgrades, my understanding is that when upgrading it uses the sources as can be seen via `choco sources`, or does it also save a source per package so it knows to fetch updates for it specifically from the source it was installed from?

**Reply by gep13** [reply]
Are you able to provide a copy of the chocolatey.log file from running the command that is failing?

No, Chocolatey CLI does not remember the source that a package was installed from.  All of the Chocolatey CLI commands will either use the sources that are configured via the `choco source` command, or use the source(s) provided via the `--source` option.

**Reply by Sainan** [reply]
Sure, but it does not contain much more information:
```
2023-05-15 13:49:47,049 27580 [DEBUG] - Resolving resource ListResource for source https://community.chocolatey.org/api/v2/
2023-05-15 13:49:47,489 27580 [DEBUG] - Resolving resource ListResource for source https://choco.calamity.gg/index.json
2023-05-15 13:49:47,540 27580 [WARN ] - The source https://choco.calamity.gg/index.json failed to get a NuGet.Protocol.Core.Types.ListResource resource
2023-05-15 13:49:47,545 27580 [INFO ] - [NuGet] Enhanced retry: HttpRetryHandler is in a state that retry would have been abandoned or not waited if it were not enabled.
2023-05-15 13:49:48,583 27580 [INFO ] - [NuGet]   GET https://community.chocolatey.org/api/v2/$metadata
2023-05-15 13:49:48,674 27580 [INFO ] - [NuGet]   OK https://community.chocolatey.org/api/v2/$metadata 90ms
2023-05-15 13:49:48,688 27580 [INFO ] - [NuGet] Enhanced retry: HttpRetryHandler is in a state that retry would have been abandoned or not waited if it were not enabled.
2023-05-15 13:49:49,828 27580 [INFO ] - [NuGet]   GET https://community.chocolatey.org/api/v2/Packages()?$filter=(tolower(Id) eq 'chocogen') and IsLatestVersion&semVerLevel=2.0.0
2023-05-15 13:49:49,916 27580 [INFO ] - [NuGet]   OK https://community.chocolatey.org/api/v2/Packages()?$filter=(tolower(Id) eq 'chocogen') and IsLatestVersion&semVerLevel=2.0.0 87ms
2023-05-15 13:49:49,918 27580 [ERROR] - chocogen not installed. The package was not found with the source(s) listed.
 Source(s): 'https://community.chocolatey.org/api/v2/;https://choco.calamity.gg/index.json'
 NOTE: When you specify explicit sources, it overrides default sources.
If the package version is a prerelease and you didn't specify `--pre`,
 the package may not be found.
Please see https://docs.chocolatey.org/en-us/troubleshooting for more
 assistance.
2023-05-15 13:49:49,924 27580 [WARN ] - 
Chocolatey installed 0/1 packages. 1 packages failed.
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log).
2023-05-15 13:49:49,924 27580 [INFO ] - 
2023-05-15 13:49:49,925 27580 [ERROR] - Failures
2023-05-15 13:49:49,925 27580 [ERROR] -  - chocogen - chocogen not installed. The package was not found with the source(s) listed.
 Source(s): 'https://community.chocolatey.org/api/v2/;https://choco.calamity.gg/index.json'
 NOTE: When you specify explicit sources, it overrides default sources.
If the package version is a prerelease and you didn't specify `--pre`,
 the package may not be found.
Please see https://docs.chocolatey.org/en-us/troubleshooting for more
 assistance.
2023-05-15 13:49:49,927 27580 [DEBUG] - Sending message 'PostRunMessage' out if there are subscribers...
2023-05-15 13:49:49,927 27580 [DEBUG] - Exiting with 1
```

However, I guess it does show you another way to repro:
```
choco install chocogen --source https://community.chocolatey.org/api/v2/;https://choco.calamity.gg/index.json
```

---

Also, where is the cache? I was trying to test the upgrade stuff with a different source but it has a faulty index.json cached, and deleting `%temp%/chocolatey` does not solve it.

**Reply by gep13** [reply]
Ok, I have been doing some looking into this, and I believe we know what is going on, but we will need to make a code change in order to fix it.

For now, you should be able to use the following:

```
choco install chocogen --source https://choco.calamity.gg/index.json
```

To install your package.

The underlying issue is that when attempting to use more than one source, if any of them have a ListResource available, it will attempt to use that, which means that the MetadataResource that is available on your V3 feed isn't queried at all.  That is why your package is not showing up when you attempt to install it.

For V3 feeds, the HTTP Cache is stored at:

```
C:\ProgramData\ChocolateyHttpCache
```

**Comment by Sainan** [comment]
Was it intended to make the 2.0.0 release (https://github.com/chocolatey/choco/releases/tag/2.0.0)?

It has no downloads and 2.0.0-beta-20230426 remains to be the latest version according to community.chocolatey.org

**Reply by pauby** [reply]
Version 2.0.0 has been released internally and will be released publicly soon.

**Comment by gep13** [comment, marked as answer]
Now that the stable releases of the Chocolatey products have shipped, I am going to close out this discussion.

If you continue to run into problems, you can join in on this discussion:

https://github.com/chocolatey/choco/discussions/3173
